### PR TITLE
makefiles/tools: Improved avrdude integration

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Currently it is impossible to use most ICSP programmers using RIOT's build system, e.g. using the `usbtiny` programmer to flash the `arduino-uno` will not work using

```
make BOARD=ardunio-uno PROGRAMMER=usbtiny
```

The reason is that the programmers serial port is set regardless of the programmer, but e.g. `usbtiny` does not use a serial port to interface with the PC. The first commit will only set the serial port for any of the default programmers of any AVR based board in RIOT (thus resulting in the same arguments for avrdude unless a custom value for `PROGRAMMER` is set). This allows using custom ICSP programmers.

The second commit sets the avrdude target based on the `$(CPU)` variable in the boards `Makefile.include`. Thus, the target no longer needs to be set for every AVR based board manually (and must no longer be set manually). The `Makefile.include` files for each board are updated accordingly.

### Testing procedure

Flashing for all AVR based boards should still work. (Instead of verifying all boards running `make BOARD=FOO flash-only` and verifying that the call to avrdude is still correct is just fine. Maybe it is good to still test for one or two boards on real hardware just to be sure.) Additionally, flashing with an ICSP (e.g. `usbtiny`) should now work.

***NOTE:*** The `jiminy-mega256rfr2` previously used `-p atmega256rfr2`, which does work. But according to the doc `-p m256rfr2` should be used. This PR changes the behavior to use `-p m256rfr2` instead.

### Issues/PRs references

None

### CODE REVIEW

**Comment by maribu** [comment]
Maybe @smlng could have a look on this?

**Comment by smlng** [comment]
on my list

File: boards/waspmote-pro/Makefile.include [hunk 1]
```diff
@@ -17,8 +17,6 @@ include $(RIOTMAKE)/tools/serial.inc.mk
 # using avrdude. Can be overridden for debugging (which requires changes
 # that require to use an ISP)
 PROGRAMMER ?= stk500v1
-# set mcu model for avrdude
-FFLAGS += -p m1281
```
Comment (jcarrano) [inline boards/waspmote-pro/Makefile.include]: D'oh! this was ugly, good that you fixed it!

**Comment by jcarrano** [comment]
I succeeded in flashing a arduino-mega2560 (uses -p). I don't have a usbtiny.

File: makefiles/tools/avrdude.inc.mk [hunk 2]
```diff
@@ -6,10 +6,16 @@ DEBUGSERVER_FLAGS = "-g -j usb :$(DEBUGSERVER_PORT)"
 DEBUGGER_FLAGS = "-x $(RIOTBOARD)/$(BOARD)/dist/gdb.conf $(ELFFILE)"
 DEBUGGER = $(DIST_PATH)/debug.sh $(DEBUGSERVER_FLAGS) $(DIST_PATH) $(DEBUGSERVER_PORT)
 
-# make the flasher port configurable (e.g. with atmelice the port is usb)
-# defaults to terminal's serial port if not configured
-AVRDUDE_PORT    ?= $(PORT)
-PROGRAMMER_FLAGS = -P $(AVRDUDE_PORT) $(FFLAGS_EXTRA)
+PROGRAMMER_FLAGS = -p $(CPU:atmega=m)
```
Comment (smlng) [inline makefiles/tools/avrdude.inc.mk]: on MacOS this doesn't seem to work, for Arduino-Mega2560 I get:

```
avrdude -c stk500v2 -p atmega2560 -P ...
```

while on master it is `m2560` bc explicitly set. HOWEVER, funnily it even works with that - but maybe that's just a coincidence and we should not rely on that.

This does not seem to be related to `make`, I tried with the default on MacOS which is version 3.81 and with one installed via HomeBrew which is 4.2.1.

Comment (maribu) [reply makefiles/tools/avrdude.inc.mk]: The full name (e.g. `atmega2560` instead of `m2560`) was actually used by the jiminy board as default. In the doc I only found the short names. But if the full name is supported officially, I would actually prefer that. I'll look that up.

If not, I think using `subst` to replace atmega by m should work anywhere.

Comment (maribu) [reply makefiles/tools/avrdude.inc.mk]: OK, the official doc does not contain any hint that the full name is supported. I consider this as an "Easter Egg" and I would favor to not rely on that.

**Comment by maribu** [comment]
@smlng: Thanks for pointing out the bug. I now changed the Makefile to use `subst` for string modification. Can you check again if it now works?

**Comment by smlng** [comment]
works as expected now, with correct `-p m<cpu>`.

File: makefiles/tools/avrdude.inc.mk [hunk 3]
```diff
@@ -6,10 +6,16 @@ DEBUGSERVER_FLAGS = "-g -j usb :$(DEBUGSERVER_PORT)"
...
-AVRDUDE_PORT    ?= $(PORT)
-PROGRAMMER_FLAGS = -P $(AVRDUDE_PORT) $(FFLAGS_EXTRA)
+PROGRAMMER_FLAGS = -p $(subst atmega,m,$(CPU))
+
+# Set flasher port only for programmers that require it
+ifneq (,$(filter $(PROGRAMMER),arduino buspirate stk500v1 stk500v2 wiring))
+  # make the flasher port configurable (e.g. with atmelice the port is usb)
+  # defaults to terminal's serial port if not configured
+  AVRDUDE_PORT    ?= $(PORT)
+  PROGRAMMER_FLAGS += -P $(AVRDUDE_PORT)
```
Comment (smlng) [inline makefiles/tools/avrdude.inc.mk]: oops wait, haven't tested that: are you sure this should be `+=`? If so please align the above `?=`

Comment (smlng) [reply makefiles/tools/avrdude.inc.mk]: looks like `+=` is good, so please realign the `?=` to make it look nice(r)

Comment (maribu) [reply makefiles/tools/avrdude.inc.mk]: Done. Shall I squash?

Comment (smlng) [reply makefiles/tools/avrdude.inc.mk]: yep, please!
